verify/verify.go

```diff
@@ -76,3 +78,51 @@ func InputVerifiedByAll(keyMappings []*KeyMapping, input []byte) (bool, []*KeyMa
 	// if the signature was verified by all of the given keys we return true
 	return len(failed) == 0, failed, nil
 }
+
+// Verify the input with given signature and a list of public keys.
+// It returns true if any of the key verifies successfuly. The file name of the successful key is returned. (true, fn, nil).
+// It returns false if all keys failed. The errors are returned as a map keyed by the key file name. The common
+// errors are keyed by COMMON_ERROR. (false, "", map).
+func InputVerifiedByAnyKey(keyFiles []string, signature string, input []byte) (bool, string, map[string]error) {
+	failed := make(map[string]error)
+
+	// no input files, failed
+	if len(keyFiles) == 0 {
+		failed[COMMON_ERROR] = fmt.Errorf("No public key files provided; input not verified")
+		return false, "", failed
+	}
+
+	// hash input
+	hasher := sha256.New()
+	_, err := hasher.Write(input)
+	if err != nil {
+		failed[COMMON_ERROR] = fmt.Errorf("Error hashing input: %v, Error: %v", string(input), err)
+		return false, "", failed
+	}
+
+	// decode the signature into its binary form.
+	signatureBytes, err := base64.StdEncoding.DecodeString(signature)
+	if err != nil {
+		failed[COMMON_ERROR] = fmt.Errorf("Unable to base64 decode signature %v, error: %v", signature, err)
+		return false, "", failed
+	}
+
+	for _, keyFile := range keyFiles {
+		// open the file
+		if pubkeyRaw, err := ioutil.ReadFile(keyFile); err != nil {
+			failed[keyFile] = fmt.Errorf("Unable to read key file: %v, error: %v", keyFile, err)
+			// check if the key is a public key
+		} else if block, _ := pem.Decode(pubkeyRaw); block == nil {
+			failed[keyFile] = fmt.Errorf("Unable to find PEM block in the provided public key: %v", keyFile)
+		} else if pubkey, err := x509.ParsePKIXPublicKey(block.Bytes); err != nil {
+			failed[keyFile] = fmt.Errorf("Unable to parse key file: %v, as a public key, error: %v", keyFile, err)
```

Comment:
This code that does the signature verification, it would be good to consolidate it across the 2 verify methods. I will leave it up to Mike if he wants you to do it now or in the future.

Reply:
Yes, let's consolidate them now. Does that work for you, Ling?